### ISSUE TITLE
Runtime panic using scalar(), again

### ISSUE THREAD

**Issue by perlun** [jacksontj/promxy#234]
Thanks for a nice tool. :+1: When testing it out, we have run into a bug which is seemingly similar to #219, but we are experiencing this with a recent version of promxy where that particular bug should be fixed.

## Description
Just like described late in #219, performing a `scalar()` query triggers an "invalid memory address or nil pointer dereference" error. Nothing gets logged to the systemd log, unfortunately.

This is an example of a query that triggers the problem. When querying Prometheus directly, the query works without issues and returns a `NaN` value.

```promql
scalar(up{instance="backend1:8083"}) * 1 +
scalar(up{instance="backend2:8083"}) * 2
```

## Versions
- Promxy v0.0.54
- Prometheus 2.13.1

## Config
```yaml
promxy:
  server_groups:
    # All upstream prometheus service discovery mechanisms are supported with the same
    # markup, all defined in https://github.com/prometheus/prometheus/blob/master/discovery/config/config.go#L33
    - static_configs:
      - targets:
        - prometheus-node1:9090
        - prometheus-node2:9090

      # anti-affinity for merging values in timeseries between hosts in the server_group
      anti_affinity: 10s

      # Controls whether to use remote_read or the prom HTTP API for fetching remote raw data
      remote_read: true
```

**Comment by jacksontj** [comment]
@perlun thanks for the report!

I attempted to reproduce locally with the query you provided but I get a `NaN` with no panic :/ 

Would you be able to provide the backtrace from the panic as well as trace level logs of the failure?

**Comment by slovdahl** [comment]
The query that triggers this error doesn't actually match anything. The query: http://prometheus-node1:8080/graph?g0.range_input=21600s&g0.end_input=2019-11-08%2008%3A12&g0.step_input=300&g0.expr=scalar(up%7Binstance%3D%22backend1%3A8083%22%7D)%20*%201%20%2B%0Ascalar(up%7Binstance%3D%22backend2%3A8083%22%7D)%20*%202&g0.tab=0

Do note that there are no metrics matching the labels in the query. I'm not that familiar with debugging Go binaries, and some quick googling didn't help me much either. Would you mind describing how to get a backtrace?

Trace log:
```
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="default scalar(up{instance=\"backend1:8083\"}) * 1 + scalar(up{instance=\"backend2:8083\"}) * 2 *promql.BinaryExpr"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="default scalar(up{instance=\"backend2:8083\"}) * 2 *promql.BinaryExpr"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="default scalar(up{instance=\"backend1:8083\"}) * 1 *promql.BinaryExpr"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="call scalar(up{instance=\"backend2:8083\"}) scalar"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="http://prometheus-node2:9090/api/v1/read" api=QueryRange query="scalar(up{instance=\"backend2:8083\"})" r="{2019-11-08 04:12:00 +0200 EET 2019-11-08 10:12:00 +0200 EET 5m0s}"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="call scalar(up{instance=\"backend1:8083\"}) scalar"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="http://prometheus-node2:9090/api/v1/read" api=QueryRange query="scalar(up{instance=\"backend1:8083\"})" r="{2019-11-08 04:12:00 +0200 EET 2019-11-08 10:12:00 +0200 EET 5m0s}"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="http://prometheus-node1:9090/api/v1/read" api=QueryRange query="scalar(up{instance=\"backend2:8083\"})" r="{2019-11-08 04:12:00 +0200 EET 2019-11-08 10:12:00 +0200 EET 5m0s}"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=debug msg="http://prometheus-node1:9090/api/v1/read" api=QueryRange query="scalar(up{instance=\"backend1:8083\"})" r="{2019-11-08 04:12:00 +0200 EET 2019-11-08 10:12:00 +0200 EET 5m0s}"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=trace msg="http://prometheus-node2:9090/api/v1/read" api=QueryRange error="<nil>" query="scalar(up{instance=\"backend2:8083\"})" r="{2019-11-08 04:12:00 +0200 EET 2019-11-08 10:12:00 +0200 EET 5m0s}" took="898.464µs" value="{} =>\nNaN @[1573179120]\nNaN @[1573179420]\nNaN @[1573179720]\nNaN @[1573180020]\nNaN @[1573180320]\nNaN @[1573180620]\nNaN @[1573180920]\nNaN @[1573181220]\nNaN @[1573181520]\nNaN @[1573181820]\nNaN @[1573182120]\nNaN @[1573182420]\nNaN @[1573182720]\nNaN @[1573183020]\nNaN @[1573183320]\nNaN @[1573183620]\nNaN @[1573183920]\nNaN @[1573184220]\nNaN @[1573184520]\nNaN @[1573184820]\nNaN @[1573185120]\nNaN @[1573185420]\nNaN @[1573185720]\nNaN @[1573186020]\nNaN @[1573186320]\nNaN @[1573186620]\nNaN @[1573186920]\nNaN @[1573187220]\nNaN @[1573187520]\nNaN @[1573187820]\nNaN @[1573188120]\nNaN @[1573188420]\nNaN @[1573188720]\nNaN @[1573189020]\nNaN @[1573189320]\nNaN @[1573189620]\nNaN @[1573189920]\nNaN @[1573190220]\nNaN @[1573190520]\nNaN @[1573190820]\nNaN @[1573191120]\nNaN @[1573191420]\nNaN @[1573191720]\nNaN @[1573192020]\nNaN @[1573192320]\nNaN @[1573192620]\nNaN @[1573192920]\nNaN @[1573193220]\nNaN @[1573193520]\nNaN @[1573193820]\nNaN @[1573194120]\nNaN @[1573194420]\nNaN @[1573194720]\nNaN @[1573195020]\nNaN @[1573195320]\nNaN @[1573195620]\nNaN @[1573195920]\nNaN @[1573196220]\nNaN @[1573196520]\nNaN @[1573196820]\nNaN @[1573197120]\nNaN @[1573197420]\nNaN @[1573197720]\nNaN @[1573198020]\nNaN @[1573198320]\nNaN @[1573198620]\nNaN @[1573198920]\nNaN @[1573199220]\nNaN @[1573199520]\nNaN @[1573199820]\nNaN @[1573200120]\nNaN @[1573200420]\nNaN @[1573200720]" warnings="[]"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=trace msg="http://prometheus-node2:9090/api/v1/read" api=QueryRange error="<nil>" query="scalar(up{instance=\"backend1:8083\"})" r="{2019-11-08 04:12:00 +0200 EET 2019-11-08 10:12:00 +0200 EET 5m0s}" took="929.228µs" value="{} =>\nNaN @[1573179120]\nNaN @[1573179420]\nNaN @[1573179720]\nNaN @[1573180020]\nNaN @[1573180320]\nNaN @[1573180620]\nNaN @[1573180920]\nNaN @[1573181220]\nNaN @[1573181520]\nNaN @[1573181820]\nNaN @[1573182120]\nNaN @[1573182420]\nNaN @[1573182720]\nNaN @[1573183020]\nNaN @[1573183320]\nNaN @[1573183620]\nNaN @[1573183920]\nNaN @[1573184220]\nNaN @[1573184520]\nNaN @[1573184820]\nNaN @[1573185120]\nNaN @[1573185420]\nNaN @[1573185720]\nNaN @[1573186020]\nNaN @[1573186320]\nNaN @[1573186620]\nNaN @[1573186920]\nNaN @[1573187220]\nNaN @[1573187520]\nNaN @[1573187820]\nNaN @[1573188120]\nNaN @[1573188420]\nNaN @[1573188720]\nNaN @[1573189020]\nNaN @[1573189320]\nNaN @[1573189620]\nNaN @[1573189920]\nNaN @[1573190220]\nNaN @[1573190520]\nNaN @[1573190820]\nNaN @[1573191120]\nNaN @[1573191420]\nNaN @[1573191720]\nNaN @[1573192020]\nNaN @[1573192320]\nNaN @[1573192620]\nNaN @[1573192920]\nNaN @[1573193220]\nNaN @[1573193520]\nNaN @[1573193820]\nNaN @[1573194120]\nNaN @[1573194420]\nNaN @[1573194720]\nNaN @[1573195020]\nNaN @[1573195320]\nNaN @[1573195620]\nNaN @[1573195920]\nNaN @[1573196220]\nNaN @[1573196520]\nNaN @[1573196820]\nNaN @[1573197120]\nNaN @[1573197420]\nNaN @[1573197720]\nNaN @[1573198020]\nNaN @[1573198320]\nNaN @[1573198620]\nNaN @[1573198920]\nNaN @[1573199220]\nNaN @[1573199520]\nNaN @[1573199820]\nNaN @[1573200120]\nNaN @[1573200420]\nNaN @[1573200720]" warnings="[]"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=trace msg="http://prometheus-node1:9090/api/v1/read" api=QueryRange error="<nil>" query="scalar(up{instance=\"backend2:8083\"})" r="{2019-11-08 04:12:00 +0200 EET 2019-11-08 10:12:00 +0200 EET 5m0s}" took=1.16307ms value="{} =>\nNaN @[1573179120]\nNaN @[1573179420]\nNaN @[1573179720]\nNaN @[1573180020]\nNaN @[1573180320]\nNaN @[1573180620]\nNaN @[1573180920]\nNaN @[1573181220]\nNaN @[1573181520]\nNaN @[1573181820]\nNaN @[1573182120]\nNaN @[1573182420]\nNaN @[1573182720]\nNaN @[1573183020]\nNaN @[1573183320]\nNaN @[1573183620]\nNaN @[1573183920]\nNaN @[1573184220]\nNaN @[1573184520]\nNaN @[1573184820]\nNaN @[1573185120]\nNaN @[1573185420]\nNaN @[1573185720]\nNaN @[1573186020]\nNaN @[1573186320]\nNaN @[1573186620]\nNaN @[1573186920]\nNaN @[1573187220]\nNaN @[1573187520]\nNaN @[1573187820]\nNaN @[1573188120]\nNaN @[1573188420]\nNaN @[1573188720]\nNaN @[1573189020]\nNaN @[1573189320]\nNaN @[1573189620]\nNaN @[1573189920]\nNaN @[1573190220]\nNaN @[1573190520]\nNaN @[1573190820]\nNaN @[1573191120]\nNaN @[1573191420]\nNaN @[1573191720]\nNaN @[1573192020]\nNaN @[1573192320]\nNaN @[1573192620]\nNaN @[1573192920]\nNaN @[1573193220]\nNaN @[1573193520]\nNaN @[1573193820]\nNaN @[1573194120]\nNaN @[1573194420]\nNaN @[1573194720]\nNaN @[1573195020]\nNaN @[1573195320]\nNaN @[1573195620]\nNaN @[1573195920]\nNaN @[1573196220]\nNaN @[1573196520]\nNaN @[1573196820]\nNaN @[1573197120]\nNaN @[1573197420]\nNaN @[1573197720]\nNaN @[1573198020]\nNaN @[1573198320]\nNaN @[1573198620]\nNaN @[1573198920]\nNaN @[1573199220]\nNaN @[1573199520]\nNaN @[1573199820]\nNaN @[1573200120]\nNaN @[1573200420]\nNaN @[1573200720]" warnings="[]"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: time="2019-11-08T10:19:04+02:00" level=trace msg="http://prometheus-node1:9090/api/v1/read" api=QueryRange error="<nil>" query="scalar(up{instance=\"backend1:8083\"})" r="{2019-11-08 04:12:00 +0200 EET 2019-11-08 10:12:00 +0200 EET 5m0s}" took=1.199697ms value="{} =>\nNaN @[1573179120]\nNaN @[1573179420]\nNaN @[1573179720]\nNaN @[1573180020]\nNaN @[1573180320]\nNaN @[1573180620]\nNaN @[1573180920]\nNaN @[1573181220]\nNaN @[1573181520]\nNaN @[1573181820]\nNaN @[1573182120]\nNaN @[1573182420]\nNaN @[1573182720]\nNaN @[1573183020]\nNaN @[1573183320]\nNaN @[1573183620]\nNaN @[1573183920]\nNaN @[1573184220]\nNaN @[1573184520]\nNaN @[1573184820]\nNaN @[1573185120]\nNaN @[1573185420]\nNaN @[1573185720]\nNaN @[1573186020]\nNaN @[1573186320]\nNaN @[1573186620]\nNaN @[1573186920]\nNaN @[1573187220]\nNaN @[1573187520]\nNaN @[1573187820]\nNaN @[1573188120]\nNaN @[1573188420]\nNaN @[1573188720]\nNaN @[1573189020]\nNaN @[1573189320]\nNaN @[1573189620]\nNaN @[1573189920]\nNaN @[1573190220]\nNaN @[1573190520]\nNaN @[1573190820]\nNaN @[1573191120]\nNaN @[1573191420]\nNaN @[1573191720]\nNaN @[1573192020]\nNaN @[1573192320]\nNaN @[1573192620]\nNaN @[1573192920]\nNaN @[1573193220]\nNaN @[1573193520]\nNaN @[1573193820]\nNaN @[1573194120]\nNaN @[1573194420]\nNaN @[1573194720]\nNaN @[1573195020]\nNaN @[1573195320]\nNaN @[1573195620]\nNaN @[1573195920]\nNaN @[1573196220]\nNaN @[1573196520]\nNaN @[1573196820]\nNaN @[1573197120]\nNaN @[1573197420]\nNaN @[1573197720]\nNaN @[1573198020]\nNaN @[1573198320]\nNaN @[1573198620]\nNaN @[1573198920]\nNaN @[1573199220]\nNaN @[1573199520]\nNaN @[1573199820]\nNaN @[1573200120]\nNaN @[1573200420]\nNaN @[1573200720]" warnings="[]"
Nov 08 10:19:04 prometheus-node1 promxy[9314]: 10.10.10.10 - - [08/Nov/2019 08:19:04] "GET /api/v1/query_range HTTP/1.1 422 126" 0.002827 query=scalar%28up%7Binstance%3D%22backend1%3A8083%22%7D%29+%2A+1+%2B%0Ascalar%28up%7Binstance%3D%22backend2%3A8083%22%7D%29+%2A+2&start=1573179120&end=1573200720&step=300&_=1573200729149
```

Running the same query directly against Prometheus I see the following raw JSON response:
```json
{"status":"success","data":{"resultType":"matrix","result":[{"metric":{},"values":[[1573179120,"NaN"],[1573179420,"NaN"],[1573179720,"NaN"],[1573180020,"NaN"],[1573180320,"NaN"],[1573180620,"NaN"],[1573180920,"NaN"],[1573181220,"NaN"],[1573181520,"NaN"],[1573181820,"NaN"],[1573182120,"NaN"],[1573182420,"NaN"],[1573182720,"NaN"],[1573183020,"NaN"],[1573183320,"NaN"],[1573183620,"NaN"],[1573183920,"NaN"],[1573184220,"NaN"],[1573184520,"NaN"],[1573184820,"NaN"],[1573185120,"NaN"],[1573185420,"NaN"],[1573185720,"NaN"],[1573186020,"NaN"],[1573186320,"NaN"],[1573186620,"NaN"],[1573186920,"NaN"],[1573187220,"NaN"],[1573187520,"NaN"],[1573187820,"NaN"],[1573188120,"NaN"],[1573188420,"NaN"],[1573188720,"NaN"],[1573189020,"NaN"],[1573189320,"NaN"],[1573189620,"NaN"],[1573189920,"NaN"],[1573190220,"NaN"],[1573190520,"NaN"],[1573190820,"NaN"],[1573191120,"NaN"],[1573191420,"NaN"],[1573191720,"NaN"],[1573192020,"NaN"],[1573192320,"NaN"],[1573192620,"NaN"],[1573192920,"NaN"],[1573193220,"NaN"],[1573193520,"NaN"],[1573193820,"NaN"],[1573194120,"NaN"],[1573194420,"NaN"],[1573194720,"NaN"],[1573195020,"NaN"],[1573195320,"NaN"],[1573195620,"NaN"],[1573195920,"NaN"],[1573196220,"NaN"],[1573196520,"NaN"],[1573196820,"NaN"],[1573197120,"NaN"],[1573197420,"NaN"],[1573197720,"NaN"],[1573198020,"NaN"],[1573198320,"NaN"],[1573198620,"NaN"],[1573198920,"NaN"],[1573199220,"NaN"],[1573199520,"NaN"],[1573199820,"NaN"],[1573200120,"NaN"],[1573200420,"NaN"],[1573200720,"NaN"]]}]}}
```

**Comment by jacksontj** [comment]
@slovdahl thanks for the data, the query you provided is also a `query_range` and with that I am able to reproduce the failure! 

I was able to repro/test with the following:
http://localhost:8082/graph?g0.range_input=1h&g0.expr=scalar(up%7Binstance%3D%22backend1%3A8083%22%7D)%20*%201%20%2B%0Ascalar(up%7Binstance%3D%22backend2%3A8083%22%7D)%20*%202&g0.tab=0&g1.range_input=1h&g1.expr=scalar(up%7Binstance%3D%22localhost%3A8000%22%2Cjob%3D%22prometheus%22%7D)&g1.tab=0&g2.range_input=1h&g2.expr=scalar(up%7Binstance%3D%22localhost%3A8000%22%2Cjob%3D%22prometheus%22%7D)&g2.tab=0

TLDR the original report was with `query` which I was unable to reproduce but @slovdahl 's report was with `query_range` which was easy to reproduce the issue. TLDR scalar is a "special" function that changes the types of things :) So #236 should finally resolve this issue once and for all. It is a bit odd that the upstream test cases didn't cover this, so that means that upstream prom doesn't have great coverage around these scalar methods :/
